### PR TITLE
[IMP] l10n_id: qr payment

### PR DESCRIPTION
**Pull request by nni-odoo**
Adding QR payment support for through QRIS

taskid-3134657

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by nni-odoo** [comment]
Additional comments:
Since we have to submit an API request to generate the content of the QR code, I'm using the following standard development as reference: https://github.com/odoo/odoo/blob/b0fbde84a444ffc4cafc13f860ad1517658af2f3/addons/l10n_in/models/account_invoice.py#L122

One improvement I could think of is: we may want to store the result of  content so that we don't have to call the API continuously when we view the invoice.

**Comment by vin-odoo** [comment]
I will take a look asap in the coming days, thanks a lot!

File: addons/l10n_id/models/account_move.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+class AccountMove(models.Model):
+    _inherit = 'account.move'
+
+    def _generate_qr_code(self, silent_errors=False):
+        """Override
+
+        Using QRIS API in order to generate the QR code
+        """
+        self.ensure_one()
+        if self.company_id.country_code == "ID":
```
Comment (vin-odoo) [inline addons/l10n_id/models/account_move.py:18]: I would also add a check on if `self.display_qr_code` is false.
Just to be safe, in the same way it is done in the original method

**Review by vin-odoo — CHANGES_REQUESTED** [review]
> we may want to store the result of content so that we don't have to call the API continuously when we view the invoice.

I think it's not that big of a deal, most of the time you'll just print and send your invoice once and that's it.
In other cases, it shouldn't hurt to do an API call from time to time.

Also, the task in the commit message isn't the correct one. Should be 3134657

Besides that and the few comments it LGTM, I tested and the QR code is generated as expected and seems valid.

**Comment by fja-odoo** [comment]
@nni-odoo @vin-odoo What if a company has 2 bank accounts that they use on different journals. With this company specific implementation they cannot do that whereas with SEPA and EMV the proxy is linked to the bank and you can have as many as you want. Isn't it possible to link multiple accounts to your qris account ?

**Comment by tong-odoo** [comment]
@nni-odoo I think it is better to add a new qr_methods in res.bank._get_available_qr_methods() to get the qr code. So the customers can choose which qr methods to show on the invoice as well

File: addons/l10n_id/models/account_move.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+            params = {
+                "do": "create-invoice",
+                "apikey": self.company_id.l10n_id_qris_apikey,
+                "mID": self.company_id.l10n_id_qris_mid,
+                "cliTrxNumber": self.payment_reference or self.name,
+                "cliTrxAmount": int(self.amount_residual)
+            }
+            res = requests.get(QRIS_API_URL, params=params).json()
```
Comment (tong-odoo) [inline addons/l10n_id/models/account_move.py:26]: Need to handle the case for timeout issue as well

Comment (nni-odoo) [reply addons/l10n_id/models/account_move.py:26]: right

**Comment by nni-odoo** [comment]
> What if a company has 2 bank accounts that they use on different journals. With this company specific implementation they cannot do that whereas with SEPA and EMV the proxy is linked to the bank and you can have as many as you want. Isn't it possible to link multiple accounts to your qris account ?
@fja-odoo from what I've gathered from HR team & JGR here that applied for QRIS, 1 QRIS account can only be linked to 1 account

**Comment by fja-odoo** [comment]
> @fja-odoo from what I've gathered from HR team & JGR here that applied for QRIS, 1 QRIS account can only be linked to 1 account

Then I think the api key and merchant id should be on the bank account instead of the company.

**Comment by nni-odoo** [comment]
> Then I think the api key and merchant id should be on the bank account instead of the company.

noted @fja-odoo

**Comment by nni-odoo** [comment]
@fja-odoo @tong-odoo @vin-odoo thanks for the input guys! I've changed the method and put all things on the res.partner.bank model as suggested.

Let me know if there's anything else!

File: addons/l10n_id/__manifest__.py [hunk 3]
```diff
@@ -13,12 +13,13 @@
     'author': 'vitraining.com',
     'website': 'https://www.odoo.com/documentation/master/applications/finance/fiscal_localizations/indonesia.html',
     'depends': [
-        'account',
+        'account_qr_code_emv',
```
Comment (tong-odoo) [inline addons/l10n_id/__manifest__.py:16]: I dont think you should depends on emv bcs you are not using emv qr code format.

**Review by tong-odoo — CHANGES_REQUESTED** [review]

File: addons/l10n_id/__manifest__.py (hunk 3 above)
Comment (nni-odoo) [reply addons/l10n_id/__manifest__.py:16]: in that case, can we move 'country_code' field of res.partner.bank from env to `base` or  `account`?

Comment (nni-odoo) [reply addons/l10n_id/__manifest__.py:16]: https://github.com/odoo/odoo/blob/910756df630c52ac143a4d25a4acd06bf2131cd5/addons/account_qr_code_emv/models/res_bank.py

Comment (tong-odoo) [reply addons/l10n_id/__manifest__.py:16]: I think its time to move it. What do you think @vin-odoo?

Comment (vin-odoo) [reply addons/l10n_id/__manifest__.py:16]: Yeah it's just a related, go for it and move this into the account module if it's needed :+1: 
@nni-odoo @tong-odoo

Comment (nni-odoo) [reply addons/l10n_id/__manifest__.py:16]: very cool thanks! @vin-odoo @tong-odoo

File: addons/l10n_id/models/res_bank.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
+import requests
```
Comment (fja-odoo) [inline addons/l10n_id/models/res_bank.py:1]: `# Part of Odoo. See LICENSE file for full copyright and licensing details.`

File: addons/l10n_id/models/res_bank.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+                "cliTrxNumber": structured_communication,
+                "cliTrxAmount": int(amount)
+            }
+            try:
+                res = requests.get(QRIS_API_URL, params=params, timeout=10)
+                res.raise_for_status()
+            except (requests.exceptions.ConnectionError, requests.exceptions.Timeout):
+                raise ValidationError(_("Could not establish the connection to the API."))
+            except requests.exceptions.HTTPError as err:
+                raise ValidationError(_("Communication wtih QRIS failed. QRIS returned the following error: %s"), err)
```
Comment (fja-odoo) [inline addons/l10n_id/models/res_bank.py:48]: with

File: addons/l10n_id/i18n/id.po [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+#. module: l10n_id
+#. odoo-python
+#: code:addons/l10n_id/models/res_bank.py:0
+msgid "Can't generate a QR code with a currency other than IDR."
+msgstr "QR Code QRIS hanya berlaku untuk mata uang Rupiah"
+
+#. module: l10n_id
+#. odoo-python
+#: code:addons/l10n_id/models/res_bank.py:0
+msgid "Communication wtih QRIS failed. QRIS returned the following error: %s"
```
Comment (fja-odoo) [inline addons/l10n_id/i18n/id.po:27]: with

**Review by fja-odoo — CHANGES_REQUESTED** [review]

File: addons/l10n_id/models/res_bank.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+            params = {
+                "do": "create-invoice",
+                "apikey": self.l10n_id_qris_api_key,
+                "mID": self.l10n_id_qris_mid,
+                "cliTrxNumber": structured_communication,
+                "cliTrxAmount": int(amount)
+            }
+            try:
+                res = requests.get(QRIS_API_URL, params=params, timeout=10)
+                res.raise_for_status()
```
Comment (vin-odoo) [inline addons/l10n_id/models/res_bank.py:44]: Please nooo, raising for status usually gives no useful information to the user besides a traceback and the status code. Could you make sure to try an error and see how it looks?
I would suggest checking by yourself and making a UserError with an ok-looking text instead to guide the user to the solution as much as possible

Comment (nni-odoo) [reply addons/l10n_id/models/res_bank.py:44]: yep I'll try some error scenarios and adjust this one

File: addons/l10n_id/i18n/l10n_id.pot
Comment (vin-odoo) [inline addons/l10n_id/i18n/l10n_id.pot:1]: for l10n modules, adding a POT file isn't useful since it won't be translated on transifex and we do translation manually in the PO file

File: odoo/addons/base/models/res_bank.py [hunk 8]
```diff
@@ -84,6 +84,7 @@ def _get_supported_account_types(self):
     sequence = fields.Integer(default=10)
     currency_id = fields.Many2one('res.currency', string='Currency')
     company_id = fields.Many2one('res.company', 'Company', related='partner_id.company_id', store=True, readonly=True)
+    country_code = fields.Char(related='partner_id.country_code', string="Country Code")
```
Comment (vin-odoo) [inline odoo/addons/base/models/res_bank.py:87]: I would challenge the need to put this in base.
It's not a big issue, but as it's only used by accounting modules I would suggest having the field inside the account module itself. (It won't change much, but we may as well keep the base module as clean as possible)

File: addons/l10n_id/models/res_bank.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    @api.model
+    def _get_available_qr_methods(self):
+        rslt = super()._get_available_qr_methods()
+        rslt.append(('id_qr', _("QRIS"), 40))
+        return rslt
+
+    def _get_error_messages_for_qr(self, qr_method, debtor_partner, currency):
+        if qr_method == 'id_qr' and self.country_code == 'ID':
+            if currency.name not in ['IDR']:
+                return _("Can't generate a QR code with a currency other than IDR.")
```
Comment (vin-odoo) [inline addons/l10n_id/models/res_bank.py:24]:
```suggestion
                return _("You cannot generate a QRIS QR code with a currency other than IDR.")
```

File: addons/l10n_id/models/res_bank.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+            if not (self.l10n_id_qris_api_key and self.l10n_id_qris_mid):
+                return _("Please configure the QRIS API Key and Merchant ID on the bank's configuration")
```
Comment (vin-odoo) [inline addons/l10n_id/models/res_bank.py:26]:
```suggestion
                return _("To use QRIS QR code, Please configure the QRIS API Key and Merchant ID on the bank's configuration")
```
Or something similar

File: addons/l10n_id/models/res_bank.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+                "do": "create-invoice",
+                "apikey": self.l10n_id_qris_api_key,
+                "mID": self.l10n_id_qris_mid,
+                "cliTrxNumber": structured_communication,
+                "cliTrxAmount": int(amount)
+            }
+            try:
+                res = requests.get(QRIS_API_URL, params=params, timeout=10)
+            except (requests.exceptions.ConnectionError, requests.exceptions.Timeout):
+                raise ValidationError(_("Could not establish the connection to the API."))
```
Comment (vin-odoo) [inline addons/l10n_id/models/res_bank.py:46]:
```suggestion
                raise ValidationError(_("Could not establish a connection to the QRIS API."))
```

File: addons/l10n_id/models/res_bank.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+            except requests.exceptions.HTTPError as err:
+                raise ValidationError(_("Communication wtih QRIS failed. QRIS returned with the following error: %s"), err)
```
Comment (vin-odoo) [inline addons/l10n_id/models/res_bank.py:48]:
```suggestion
                raise ValidationError(_("Communication with QRIS failed. QRIS returned with the following error: %s"), err)
```

**Review by vin-odoo — CHANGES_REQUESTED** [review]
I see on the documentation that the QR code is only valid for 30m. Isn't this an issue?
I would expect that a company will quite often not pay their invoices in the 30m following their reception.

File: odoo/addons/base/models/res_bank.py (hunk 8 above)
Comment (nni-odoo) [reply odoo/addons/base/models/res_bank.py:87]: noted on this, the reason for is because `l10n_hk`, `l10n_sg`, `l10n_vn` etc. , res_bank_views.xml are all inheriting from 'base.view_partner_bank_form'. If we put it on account, then I would need to change the inherit_id for all these views that use country_code. It was only for convenience :grin: 
But noted, I'll move it to account

Comment (vin-odoo) [reply odoo/addons/base/models/res_bank.py:87]: Hmm maybe base is ok then... It's not better to go change inherits everywhere

Comment (nni-odoo) [reply odoo/addons/base/models/res_bank.py:87]: noted!

**Comment by nni-odoo** [comment]
> I see on the documentation that the QR code is only valid for 30m. Isn't this an issue? I would expect that a company will quite often not pay their invoices in the 30m following their reception.

Yes you're right it's only valid for 30 minutes. From the doc, we can pass an extra argument: "qris_request_date" to help set the expiry date.  I suppose we we can set the request date to a certain number here (i.e. like 1 week)?
From a practical POV, if the customer decides to pay, wouldn't they go to the portal page which then will trigger the creation of QR code at that point. So, I'm assuming in practice it would still be safe?

Let me know if you have other suggestions

**Comment by vin-odoo** [comment]
> > I see on the documentation that the QR code is only valid for 30m. Isn't this an issue? I would expect that a company will quite often not pay their invoices in the 30m following their reception.
> 
> Yes you're right it's only valid for 30 minutes. From the doc, we can pass an extra argument: "qris_request_date" to help set the expiry date. I suppose we we can set the request date to a certain number here (i.e. like 1 week)? From a practical POV, if the customer decides to pay, wouldn't they go to the portal page which then will trigger the creation of QR code at that point. So, I'm assuming in practice it would still be safe?
> 
> Let me know if you have other suggestions

From my understanding, qris_request_date will not impact the expiry. It will expire 30m after its **generation**.

> From a practical POV, if the customer decides to pay, wouldn't they go to the portal page which then will trigger the creation of QR code at that point. 

I'm not sure, I would just open the received invoice from my email and pay from there.

**Comment by nni-odoo** [comment]
> From my understanding, qris_request_date will not impact the expiry. It will expire 30m after its generation.

Based on the following image, I think the definition is saying that we can utilize `qris_request_date` to control the expiry. I can try to see if it really works.
image

**Comment by nni-odoo** [comment]
@vin-odoo I'll ask the QRIS team to see if they have any alternative to this. In the case that they don't provide any methods to prolong or extendt he expiry of qr code, do we have an alternative to it?

**Comment by nni-odoo** [comment]
@vin-odoo as we discussed yesterday, I talked to QRIS team and they mentioned there's no way for us to extend the expiry date of QRIS.
Per our discussion as well:
I've added new method on account.move called _get_qr_additional_note which will then be shown on the invoice report. 
Let me know if there's anything else!

**Comment by nni-odoo** [comment]
@fja-odoo @vin-odoo adjusted the code so that QR code is only displayed when accessing from portal page.

Let me know if there's anything else

File: addons/l10n_id/models/res_bank.py [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+                raise ValidationError(_("Communication with QRIS failed. QRIS returned with the following error: %s", err_msg))
+            qris_content = response.get("data").get("qris_content")
+            return qris_content
+
+        return super()._get_qr_vals(qr_method, amount, currency, debtor_partner, free_communication, structured_communication)
+
+    def _get_qr_code_generation_params(self, qr_method, amount, currency, debtor_partner, free_communication, structured_communication):
+        if qr_method == 'id_qr':
+            if not self._context.get('from_portal'):
+                return None
```
Comment (vin-odoo) [inline addons/l10n_id/models/res_bank.py:65]:
```suggestion
                return {}
```
Just to keep a single return type

**Review by vin-odoo — CHANGES_REQUESTED** [review]
Forgot to send the review, sorry for the delay

File: addons/l10n_id/models/res_bank.py [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+        return super()._get_error_messages_for_qr(qr_method, debtor_partner, currency)
+
+    def _get_qr_vals(self, qr_method, amount, currency, debtor_partner, free_communication, structured_communication):
+        """Override
+
+        Getting content for the QR through calling QRIS API"""
+        if qr_method == "id_qr":
+            # if it's coming from email and is QR, then return none immediately
+            if self._context.get('default_mail_template_id'):
+                return None
```
Comment (vin-odoo) [inline addons/l10n_id/models/res_bank.py:38]: Still there
image

File: addons/l10n_id/controllers/portal.py [hunk 15]
```diff
@@ -0,0 +1,10 @@
+
+from odoo.addons.account.controllers.portal import PortalAccount
+from odoo import http, _
+from odoo.http import request
+
+class Portal(PortalAccount):
+    @http.route()
+    def portal_my_invoice_detail(self, **kw):
+        request.env.context = {**request.env.context, "from_portal": True}
```
Comment (vin-odoo) [inline addons/l10n_id/controllers/portal.py:9]: A small comment to explain the need of that context key would be welcome :)

File: addons/l10n_id/models/res_bank.py (hunk 14 above)
Comment (nni-odoo) [reply addons/l10n_id/models/res_bank.py:38]: wait what

File: addons/l10n_id/models/res_bank.py [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+QRIS_API_URL = "https://qris.online/restapi/qris/show_qris.php"
+
+class ResBank(models.Model):
+    _inherit = "res.partner.bank"
+
+    l10n_id_qris_api_key = fields.Char("QRIS API Key")
+    l10n_id_qris_mid = fields.Char("QRIS Merchant ID")
```
Comment (vin-odoo) [inline addons/l10n_id/models/res_bank.py:13]: We'll need to reevaluate the other module (emv) on this, but I think this is not very safe and goes against the safety mechanism we put in place last time (see allow_out_payment)

tldr; fields should be readonly when allow_out_payment is true. this should be enforced in the write, similar to what's done in res_partner_bank in the account module

**Comment by vin-odoo** [comment]
@odoo/rd-security Hello team, could we get a check, please?

Request is used to do an API request to get the QR code from QRIS. It is done when displaying the invoice PDF in the portal, but not in other cases.

File: addons/l10n_id/models/res_bank.py [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+            return None
+
+        return super()._get_error_messages_for_qr(qr_method, debtor_partner, currency)
+
+    def _get_qr_vals(self, qr_method, amount, currency, debtor_partner, free_communication, structured_communication):
+        """Override
+
+        Getting content for the QR through calling QRIS API"""
+        if qr_method == "id_qr":
+            # if it's coming from email and is QR, then return none immediately
```
Comment (vin-odoo) [inline addons/l10n_id/models/res_bank.py:36]:
```suggestion
```

**Review by vin-odoo — APPROVED** [review]

**Comment by vin-odoo** [comment]
Note for the other reviewers: The related country_code is moved to avoid repeating the field multiple time.
as it's a very simple non-stored related, I believe it shouldn't be an issue?

File: addons/l10n_id/i18n/id.po [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+msgid ""
+"Communication with QRIS failed. QRIS returned with the following error: %s"
+msgstr "Komunikasi dengan QRIS gagal. QRIS kembali dengan error berikut: %s"
+
+#. module: l10n_id
+#. odoo-python
+#: code:addons/l10n_id/models/res_bank.py:0
+msgid "Could not establish a connection to the QRIS API."
+msgstr "Gagal membuat sambungan ke API"
+
```
Comment (mart-e) [inline addons/l10n_id/i18n/id.po:31]: empty line. Moreover this file is in `i18n` and your remove pot in `i18n_extra`, you can move the pot here instead

**Comment by mart-e** [comment]
@robodoo override=ci/security
Reason: requests to API only

File: addons/l10n_id/i18n/id.po (hunk 18 above)
Comment (vin-odoo) [reply addons/l10n_id/i18n/id.po:31]: I think pot is not needed in l10n modules? As they're not used anyway

Thanks for the override

Comment (mart-e) [reply addons/l10n_id/i18n/id.po:31]: Why do you say they are not used? We use .pot files when reading po
https://github.com/odoo/odoo/blob/17.0/odoo/tools/translate.py#L639-L651
